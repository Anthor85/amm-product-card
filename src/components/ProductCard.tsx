import React, { createContext, JSX } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  value?: number;
  onChange?: (args: onChangeArgs) => void;
  // children?: ReactElement | ReactElement[];
  children: (props: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  value,
  onChange,
  className = '',
  style,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    isMaxCountReached,
    maxCount,
    reset,
  } = useProduct({
    product,
    value,
    initialValues,
    onChange,
  });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div
        className={`${styles.productCard} ${className}`}
        key={product.id}
        style={style}
      >
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
