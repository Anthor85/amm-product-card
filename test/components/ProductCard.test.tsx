import React from 'react';
import render from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product1 } from '../data/products';

const { act } = render;

describe('ProductCard', () => {
  test('debe mostrar el componente correctamente', () => {
    const wrapper = render.create(
      <ProductCard product={product1}>{() => <h1>Card</h1>}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe incrementar el contador', () => {
    const wrapper = render.create(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>ProductCard</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}>+</button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    act(() => {
      const button = (tree as any).children?.[2];
      button.props.onClick();
    });
    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
