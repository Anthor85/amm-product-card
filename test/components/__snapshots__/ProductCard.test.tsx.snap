// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard debe incrementar el contador 1`] = `
<div
  className="productCard "
>
  <h1>
    ProductCard
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;

exports[`ProductCard debe mostrar el componente correctamente 1`] = `
<div
  className="productCard "
>
  <h1>
    Card
  </h1>
</div>
`;
